### DISCUSSION TITLE
How to test the ECC function of E51 and U54?

### DISCUSSION THREAD

**Discussion by Balle0500** [polarfire-soc/polarfire-soc-documentation#404] [HSS and U-Boot] [answered]
May I ask if ECC for L1 cache, L2 cache, and DDR is enabled by default? Can ECC be turned off? How to test if the ECC function is effective?

**Comment by hughbreslin** [comment]
Hi @Balle0500 have you seen this [document](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/fundamentals/ecc/handling-ecc-errors-on-polarfire-soc.md) on ECC? There are ECC error injection registers to test ECC and the HSS can print out ECC stats from the [latest release](https://github.com/polarfire-soc/hart-software-services/releases). The addresses for the individual registers and documentation on how to use them can be found in the [TRM](https://ww1.microchip.com/downloads/aemDocuments/documents/FPGA/ProductDocuments/ReferenceManuals/PolarFire_SoC_FPGA_MSS_Technical_Reference_Manual_VC.pdf) or the register map. 

ECC is there for L1 and L2 by default, if you have DDR that supports ECC it is available and the implementation is explained in the first document I linked. I don't think you can turn off ECC but you can mask interrupts so theres notification of an error.

**Reply by Balle0500** [reply]
Thank you for your reply! May I ask how to use  ECC error injection registers? I didn't see any relevant explanation about them in the document. Also, how can I test whether the ECC interrupt function of L2 cache works well?
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/155931587/40f8fc9a-da56-4285-9294-4e2f3143b379)
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/155931587/a37fdb28-a3de-496d-8257-b469fd8dbc55)

**Reply by hughbreslin** [reply, marked as answer]
There is a description in the TRM:
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/44862881/e1c7efb3-e518-4f1d-8a81-729d9af64582)

You just need to set which bit(s) you would like to toggle and the error injection is enabled :)

In terms of testing, you will need to enable and configure interrupts (the MSS GPIO bare metal example is useful for this), set the error injection register, write to the L2 and then readback from the L2. In terms of the cache you can set the error injection bits but you will need to wait for writes and readbacks on the cache which you can't directly control. Your best bet would be to write to DDR via the cache and then readback.
